BParkingNano/plugins/TriggeringMuonProducer.cc

```diff
@@ -0,0 +1,173 @@
+}
+
+
+
+void TriggeringMuonProducer::produce(edm::Event& iEvent, const edm::EventSetup& iSetup) {
+    
+
+    edm::Handle<reco::VertexCollection> vertexHandle;
+    iEvent.getByToken(vertexSrc_, vertexHandle);
+    const reco::Vertex & PV = vertexHandle->front();
+
+    if(debug) std::cout << " BToKstllProducer::identifyTriggeringMuons " << std::endl;
+
+    edm::Handle<edm::TriggerResults> triggerBits;
+    iEvent.getByToken(triggerBits_, triggerBits);
+    const edm::TriggerNames &names = iEvent.triggerNames(*triggerBits);
+
+    std::vector<std::vector<float>> triggeringMuons;
+    //taken from https://twiki.cern.ch/twiki/bin/view/CMSPublic/WorkBookMiniAOD2016#Trigger
+    edm::Handle<std::vector<pat::TriggerObjectStandAlone>> triggerObjects;
+    iEvent.getByToken(triggerObjects_, triggerObjects);
+    if(debug) std::cout << "\n TRIGGER OBJECTS " << std::endl;
+
+    for (pat::TriggerObjectStandAlone obj : *triggerObjects) { // note: not "const &" since we want to call unpackPathNames
+      obj.unpackFilterLabels(iEvent, *triggerBits);
+      obj.unpackPathNames(names);
+
+      bool isTriggerMuon = false;
+      for (unsigned h = 0; h < obj.filterIds().size(); ++h)
+	if(obj.filterIds()[h] == 83){ 
+	  isTriggerMuon = true; 
+	  if(debug) std::cout << "\t   Type IDs:   " << 83;  //83 = muon
+	  break;
+	} 
+
+      if(!isTriggerMuon) continue; 
+      for (unsigned h = 0; h < obj.filterLabels().size(); ++h){
+	std::string filterName = obj.filterLabels()[h];
+	if(filterName.find("hltL3") != std::string::npos  && filterName.find("Park") != std::string::npos){
+	  isTriggerMuon = true;
+	  if(debug) std::cout << "\t   Filters:   " << filterName; 
+	  break;
+	}
+	else{ isTriggerMuon = false; }
+      }
+
+      if(!isTriggerMuon) continue;
+      std::vector<float> localMuon;
+      localMuon.push_back(obj.pt());
+      localMuon.push_back(obj.eta());
+      localMuon.push_back(obj.phi());
+      triggeringMuons.push_back(localMuon);
+      if(debug){ std::cout << "\tTrigger object:  pt " << obj.pt() << ", eta " << obj.eta() << ", phi " << obj.phi() << std::endl;
+	// Print trigger object collection and type
+	std::cout << "\t   Collection: " << obj.collection() << std::endl;
+      }
+    }//trigger objects
+
+    if(debug){
+      std::cout << "\n total n of triggering muons = " << triggeringMuons.size() << std::endl;
+      for(auto ij : triggeringMuons){
+	for(auto il : ij) std::cout << " >>> components (pt, eta, phi) = " << il << std::endl;
+      }
+    }
+
+
+    std::unique_ptr<pat::CompositeCandidateCollection> result( new pat::CompositeCandidateCollection );
+
+
+    //now check for reco muons matched to triggering muons
+    edm::Handle<std::vector<pat::Muon>> muonForTrgHandle;
+    iEvent.getByToken(muonSrc_, muonForTrgHandle);
+
+    for(unsigned int iTrg=0; iTrg<muonForTrgHandle->size(); ++iTrg){
+      const pat::Muon & muon1 = (*muonForTrgHandle)[iTrg];
+
+      if(!(muon1.isLooseMuon() && muon1.isSoftMuon(PV))) continue;
```

Comment:
Please avoid hardcoded selections, we can put some upstream in a dedicated selector.

Reply:
this is for the trigger muon. I can think a way to use a string with the cut in the configuration. But it's low priority for the moment (we will not change the requirements on the reco muons matched to the trigger...)